### DISCUSSION TITLE
Managing User Activation Status and Database Operations in Multi-Role Project

### DISCUSSION THREAD

**Discussion by nisarg2023** [supabase/supabase#19370] [Questions]
I have developed a project involving multiple roles such as admin, super admin, and user. Super admin privileges include the ability to activate or deactivate users within their portal. Currently, I have implemented a mechanism to check a user's activation status at login; however, I am facing an issue when a user is already logged in and the super admin subsequently deactivates them.

My objective is to restrict inactive users from performing any operations on the database. In my project, a client component is utilized. I am seeking guidance on how to gracefully log out a user in such scenarios to prevent them from executing any further operations on the database.

**Comment by GaryAustin1** [comment]
With jwt's (and I assume you are using the jwt to store a role claim, but you don't say), the jwt is good until it expires.  The automatic refresh would load the new jwt or you can force a refresh with refreshToken from the client.   There is no way from the Supabase server to force this.  There is no check on the jwt on every transaction.   If your roles were in a table and  you checked that in RLS on every operation then that would immediately be reflected.

**Reply by GaryAustin1** [reply]
I just started work on a "demo" of how to do this with a table instead of the jwt.  The results as far as performance look good.
https://github.com/GaryAustin1/custom-properties .

**Reply by nisarg2023** [reply]
@GaryAustin1  can I use the edge function? instead of RLS

**Reply by nisarg2023** [reply]
![image](https://github.com/supabase/supabase/assets/125979189/8d1bc534-502b-4fc1-b099-4fe706e185fa)

if I use RLS it does not give the error  message if a user is not Active
so how do I manage User Activation Status any alternative solution available?

**Reply by nisarg2023** [reply]
I am currently considering two approaches for managing user access to data and operations within our system. The first option involves implementing Row-Level Security (RLS) for each table, where the user's status would determine their access privileges, enabling or restricting actions such as update, insert, or delete. The second option entails manually querying the user status before permitting access and performing operations. I am seeking guidance on which approach is considered the best practice for ensuring secure and controlled data access within our system. Your insights would be greatly appreciated."

**Reply by GaryAustin1** [reply]
If the 2nd approach involves your own server code AND you have RLS disabled for authenticated and anon users (or tables not granted to them with Postgres grants).  Then that is a fine way to go.  Otherwise using RLS and either the established custom-claims approach using the jwt or a table approach like I showed with WELL WRITTEN RLS is also fine.    If your server code has to do an extra fetch to the server then in many cases the RLS approach will be faster.

**Comment by nisarg2023** [comment]
Thank you, @GaryAustin1, for your response. 
Could you please inform me when the demonstration is complete?
